### PR TITLE
Implement builds of Docker images (closes #1141)

### LINKED ISSUE #1141: new feature - docker
It would be great to run testcafe tests inside docker, to allow executing on a docker hosts with other tests.

Now there is requirement to launch testcafe on a system with locally installed browsers.

Would be nice to have some predefined docker images with different browsers to launch tests without installing chrome\firefox etc on a server.

### PR DESCRIPTION
**Pull request by AndreyBelym**
\cc @inikulin @AlexanderMoskovkin 
Hub for our images: https://hub.docker.com/r/testcafe/testcafe/

Container comes with Chromium and Firefox preinstalled with Xvfb, so it allows headless testing out-of-the-box.

## Build How-To For Windows
Get the [Docker Toolbox](https://github.com/docker/toolbox/releases) and install it.
Open Docker Quickstart Terminal and navigate to the TestCafe repository.
Do `gulp docker-build` or `gulp docker-publish`. If you are planning to publish the image, do `docker login` with TestCafe credentials before running `gulp docker-publish`.
Image will be published with the dist tag specified in the `.publishrc`.
## How To Use Example
`docker pull testcafe/testcafe:alpha`
`docker run -v ${TEST_FOLDER}:/tests -it testcafe/testcafe:alpha 'chromium --no-sandbox,firefox' /tests/test.js`, e. g.
`docker run -v //c/Users/belym.andrey/tests:/tests -it testcafe/testcafe:alpha 'chromium --no-sandbox,firefox' /tests/test.js`

Actually, IMHO we need a recipe for that because there is a lot of caveats with docker, especially on Windows. https://docs.docker.com/engine/tutorials/dockervolumes/#/mount-a-host-directory-as-a-data-volume

### CODE REVIEW

**Comment by inikulin** [comment]
I guess it would be nice to publish them automatically as post-publish script for publish-please. publish-please don't have these functionality yet though. It should be very easy to implement. @AndreyBelym would like to issue a PR to publish-please?

**Comment by inikulin** [comment]
ping @AndreyBelym

**Comment by AndreyBelym** [comment]
OK, I will do a PR to `publish-please`.

**Comment by YamiOdymel** [comment]
Error occurred while using `testcafe` with `chromium` with the docker image.

**command**

```bash
node /opt/testcafe/bin/testcafe.js chromium --no-sandbox,firefox ./test.js
```

**output**

```
ERROR Unable to establish one or more of the specified browser connections. This can be caused by network issues or remote device failure.
```

---

But `testcafe` works fine If I left `firefox` only.

**command**

```bash
node /opt/testcafe/bin/testcafe.js firefox ./test.js
```

**output**

```
Running tests in:
35s
407
 - Firefox 45.0.0 / Linux 0.0.0
35s
409
 Example page
35s
410
 ✓ Emulate user actions and perform a verification
39s
413
 1 passed (4s)
```

---

**test.js**

```js
fixture `Example page`
    .page `https://devexpress.github.io/testcafe/example`;

test('Emulate user actions and perform a verification', async t => {
    await t
        .setNativeDialogHandler(() => true)
        .click('#populate')
        .click('#submit-button');

    const location = await t.eval(() => window.location);

    await t.expect(location.pathname).eql('/testcafe/example/thank-you.html');
});
```

**Comment by jameswlane** [comment]
First off 👍 on Docker

How are you planning to be able to pass in other flags?
For example the app flags.
ie: `testcafe nightmare tests/**/*.js --app \"node server.js\" --app-init-delay 4000`

**Comment by AndreyBelym** [comment]
Hello @jameswlane! Yeah, the image should accept any arguments that you can specify for the regular TestCafe CLI. But currently you can only use Chrome and Firefox that are built into the image, not the Nightmare.

**Comment by jameswlane** [comment]
Awesome!

I been trying to get it to run and I keep getting the following errors:
`docker run -v $PWD/tests:/tests -it testcafe/testcafe:alpha 'firefox' /tests/**/*.js --app "node server.js" --app-init-delay 4000`

```
ERROR Tested app failed with an error:

Error: Command failed: node server.js
module.js:471
    throw err;
    ^

Error: Cannot find module '/server.js'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3

    at ChildProcess.exithandler (child_process.js:206:12)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:877:16)
    at Socket.<anonymous> (internal/child_process.js:334:11)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at Pipe._handle.close [as _onclose] (net.js:498:12)

Type "testcafe -h" for help.
error Command failed with exit code 1.
```

I assumed it was only mapping the test directories so I tried:
`docker run -v $PWD:/tests -it testcafe/testcafe:alpha 'firefox' /tests/**/*.js --app "node server.js" --app-init-delay 4000 `

```
ERROR Cannot prepare tests due to an error.

Error: Cannot find module 'tap'
    at Object.<anonymous> (/tests/node_modules/fast-safe-stringify/test.js:1:12)

Type "testcafe -h" for help.
error Command failed with exit code 1.
```

**Comment by AndreyBelym** [comment]
If you want to use `--app 'node server.js``, you should map the directory with `server.js` into the TestCafe volume. About `Error: Cannot find module 'tap'`: are you sure that you've installed the `tap` module in the `${PWD}/node_modules`?

**Comment by AndreyBelym** [comment]
Hello @YamiOdymel! Quite strange, I just ran it successfully with the command from the PR starting post:
```
docker run -v //c/Users/belym.andrey/tests:/tests -it testcafe/testcafe:alpha 'chromium --no-sandbox,firefox' /tests/test.js
```
How did you exactly run the image?

**Comment by jameswlane** [comment]
@AndreyBelym 

Here is the [codebase](https://github.com/jameswlane/example-hapi-testcafe-acceptance-testing) I am testing with. 

When you run `npm test` / `yarn test`:
```
λ yarn test   
yarn test v0.21.3
$ testcafe nightmare tests/**/*.js --app "node server.js" --app-init-delay 4000 
 Running tests in:
 - Electron 1.6.3 / Mac OS X 10.11.6

 Getting Started
 ✓ Check property of element

 Getting Started
 ✓ Check property of element


 2 passed (1s)
✨  Done in 11.56s.
```

When you run `npm test:ci` / `yarn test:ci`:
```
yarn test:ci v0.21.3
$ docker run -v $PWD:/tests -it testcafe/testcafe:alpha 'firefox' /tests/**/*.js --app "node server.js" --app-init-delay 4000 
ERROR Cannot prepare tests due to an error.

Error: Cannot find module 'tap'
    at Object.<anonymous> (/tests/node_modules/fast-safe-stringify/test.js:1:12)

Type "testcafe -h" for help.
error Command failed with exit code 1.
```

So I am think I have the directory mapped correctly but I do not have or use `tap` as a dependency.

As a experiment I installed tap and ran `npm test:ci` / `yarn test:ci`:
```
yarn test:ci
yarn test:ci v0.21.3
$ docker run -v $PWD:/tests -it testcafe/testcafe:alpha 'firefox' /tests/**/*.js --app "node server.js" --app-init-delay 4000 
⠦
TAP version 13
# Subtest: circular reference to root
    ok 1 - should be equal
    1..1
ok 1 - circular reference to root # time=5.911ms

# Subtest: nested circular reference to root
    ok 1 - should be equal
    1..1
ok 2 - nested circular reference to root # time=1.318ms

# Subtest: child circular reference
    ok 1 - should be equal
    1..1
ok 3 - child circular reference # time=0.645ms

# Subtest: nested child circular reference
    ok 1 - should be equal
    1..1
ok 4 - nested child circular reference # time=1.193ms

# Subtest: circular objects in an array
    ok 1 - should be equal
    1..1
ok 5 - circular objects in an array # time=0.905ms

# Subtest: nested circular references in an array
    ok 1 - should be equal
    1..1
ok 6 - nested circular references in an array # time=0.616ms

# Subtest: circular arrays
    ok 1 - should be equal
    1..1
ok 7 - circular arrays # time=0.598ms

# Subtest: nested circular arrays
    ok 1 - should be equal
    1..1
ok 8 - nested circular arrays # time=0.653ms

# Subtest: repeated non-circular references in objects
    ok 1 - should be equal
    1..1
ok 9 - repeated non-circular references in objects # time=0.858ms

# Subtest: repeated non-circular references in arrays
    ok 1 - should be equal
    1..1
ok 10 - repeated non-circular references in arrays # time=0.865ms

# Subtest: double child circular reference
    ok 1 - should be equal
    ok 2 - should be equivalent
    1..2
ok 11 - double child circular reference # time=3.685ms
ERROR Cannot prepare tests due to an error.

TypeError: Cannot read property 'Date' of undefined
    at /tests/node_modules/moment-duration-format/test/vendor/qunit.js:22:15
    at Object.<anonymous> (/tests/node_modules/moment-duration-format/test/vendor/qunit.js:2212:3)

Type "testcafe -h" for help.
1..11
# time=2471.47ms
error Command failed with exit code 1.
```

**Comment by AndreyBelym** [comment]
OK, I figured out, you mapped the entire project directory as tests and tried to use `/tests/**/*.js` as the wildcard for test files, and TestCafe started to execute every `js`-file from your project directoy, including ones from `node_modules`. I guess you can try
```
docker run -v $PWD:/project -it testcafe/testcafe:alpha 'firefox' /project/tests/**/*.js --app "node /project/server.js" --app-init-delay 4000
```

**Comment by jameswlane** [comment]
@AndreyBelym - You are awesome! 👍 

```
λ yarn test:ci
yarn test:ci v0.21.3
$ docker run -v $PWD:/project -it testcafe/testcafe:alpha 'firefox' /project/tests/**/*.js --app "node /project/server.js" --app-init-delay 4000 
 Running tests in:
 - Firefox 45.0.0 / Linux 0.0.0

 Getting Started
 ✓ Check property of element

 Getting Started
 ✓ Check property of element


 2 passed (2s)
✨  Done in 13.17s.
```

File: Gulpfile.js [hunk 1]
```diff
@@ -578,15 +578,57 @@ gulp.task('test-functional-travis-old-browsers', ['build'], function () {
...
+
+    var dockerEnv = getDockerEnv(dockerMachineName);
+
+    assignIn(process.env, dockerEnv);
+
+    return true;
+}
+
 gulp.task('docker-build', function () {
+    if (!process.env['DOCKER_HOST'] && !startDocker())
```
Comment (inikulin) [inline Gulpfile.js]: `startDocker` always returns `true`. What the purpose of this condition?

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

**Pull request merged**